### PR TITLE
Shared model/new consensus gate

### PR DESCRIPTION
**Pull request by x3medima17**
### Description of the Change
Migrate consensus gate to the new model

### Benefits
Less dependency on old model

### Possible Drawbacks 
None

### CODE REVIEW

File: irohad/consensus/yac/impl/yac_gate_impl.cpp [hunk 1]
```diff
@@ -42,27 +49,36 @@ namespace iroha {
             block_loader_(std::move(block_loader)),
             delay_(delay) {
         log_ = logger::log("YacGate");
-        block_creator_->on_block().subscribe(
-            [this](auto block) { this->vote(block); });
+        block_creator_->on_block().subscribe([this](auto block) {
+          auto bl = shared_model::proto::from_old(block);
+          auto tmp = bl.prevHash();
```
Comment (l4l) [inline irohad/consensus/yac/impl/yac_gate_impl.cpp]: looks unused

File: irohad/consensus/yac/impl/yac_gate_impl.cpp [hunk 2]
```diff
@@ -42,27 +49,36 @@ namespace iroha {
...
+          this->vote(bl);
+        });
       }
 
-      void YacGateImpl::vote(model::Block block) {
-        auto hash = hash_provider_->makeHash(block);
+      void YacGateImpl::vote(const shared_model::interface::Block &block) {
+        std::cout << block.signatures().size() << std::endl;
```
Comment (l4l) [inline irohad/consensus/yac/impl/yac_gate_impl.cpp]: shouldn't be here

File: irohad/validation/impl/chain_validator_impl.cpp [hunk 3]
```diff
@@ -37,7 +37,7 @@ namespace iroha {
                  block.height(),
                  block.hash().hex());
       auto apply_block =
-          [this](const auto &block, auto &queries, const auto &top_hash) {
+          [&](const auto &block, auto &queries, const auto &top_hash) {
```
Comment (l4l) [inline irohad/validation/impl/chain_validator_impl.cpp]: why?

File: shared_model/backend/protobuf/block.hpp [hunk 4]
```diff
@@ -80,6 +80,11 @@ namespace shared_model {
         return true;
       }
 
+      bool clearSignatures() override {
+        signatures_->clear();
+        return (signatures_->size() == 0);
```
Comment (l4l) [inline shared_model/backend/protobuf/block.hpp]: Is there any case when it can return false?
P.s: If you're considering multi-thread behavior, e.g: here goes cleaning and somewhere else Lazy field invalidated, then there's still an rc-bug may happen. Moreover I don't think there's an "easy" solution of that problem with the current design

File: shared_model/builders/protobuf/builder_templates/block_template.hpp [hunk 5]
```diff
@@ -53,7 +53,6 @@ namespace shared_model {
 
       enum RequiredFields {
         Transactions,
-        TxNumber,
```
Comment (l4l) [inline shared_model/builders/protobuf/builder_templates/block_template.hpp]: Looks unrelated to the pr

File: test/integration/consensus/consensus_sunny_day.cpp [hunk 6]
```diff
@@ -29,10 +29,9 @@ using namespace iroha::model;
 using namespace iroha::consensus::yac;
 using namespace framework::test_subscriber;
 
-Peer mk_local_peer(uint64_t num) {
-  Peer peer;
-  peer.address = "0.0.0.0:" + std::to_string(num);
-  return peer;
+auto mk_local_peer(uint64_t num) {
+  auto address = "0.0.0.0:" + std::to_string(num);
```
Comment (l4l) [inline test/integration/consensus/consensus_sunny_day.cpp]: D: why "0.0.0.0" is considered a local peer
Replace with 127.0.0.1 plz, also I think you may inline this var

File: test/module/irohad/consensus/yac/cluster_order_test.cpp [hunk 7]
```diff
@@ -17,20 +17,21 @@
...
 #include "consensus/yac/cluster_order.hpp"
 #include <gtest/gtest.h>
+#include "module/irohad/consensus/yac/yac_mocks.hpp"
 
 class ClusterOrderTest : public ::testing::Test {
  protected:
   void SetUp() override {
-    p1.address = "1";
-    p2.address = "2";
+    p1 = iroha::consensus::yac::mk_peer("1");
```
Comment (l4l) [inline test/module/irohad/consensus/yac/cluster_order_test.cpp]: move initialization to the var declaration

File: test/module/irohad/consensus/yac/yac_gate_test.cpp [hunk 8]
```diff
@@ -112,8 +115,14 @@ TEST_F(YacGateTest, YacGateSubscriptionTest) {
...
   // verify that yac gate emit expected block
   auto gate_wrapper = make_test_subscriber<CallExact>(gate->on_commit(), 1);
-  gate_wrapper.subscribe(
-      [this](auto block) { ASSERT_EQ(block, expected_block); });
+  gate_wrapper.subscribe([this](auto block) {
+    auto bl = *std::unique_ptr<iroha::model::Block>(block->makeOldModel());
+    // std::cout<<block->toString()<<std::endl;
+    std::cout << shared_model::proto::from_old(expected_block).hash().hex()
+              << std::endl;
+    std::cout << expected_block.hash.to_hexstring() << std::endl;
```
Comment (l4l) [inline test/module/irohad/consensus/yac/yac_gate_test.cpp]: Remove cout-s plz

File: test/module/irohad/consensus/yac/yac_mocks.hpp [hunk 9]
```diff
@@ -215,13 +226,15 @@ namespace iroha {
...
         // ------|Round|------
-        std::vector<model::Peer> default_peers = [] {
-          std::vector<model::Peer> result;
-          for (size_t i = 1; i <= 7; ++i) {
-            result.push_back(mk_peer(std::to_string(i)));
-          }
-          return result;
-        }();
+        std::vector<std::shared_ptr<shared_model::interface::Peer>>
+            default_peers = [] {
```
Comment (l4l) [inline test/module/irohad/consensus/yac/yac_mocks.hpp]: Replace with std::generate_n plz, will look a bit prettier:

```cpp
std::vector<...> default_peers;
std::generate_n(default_peers.begin(), default_peers.end(), 7, [i = 1] mutable {
  return mk_peer(std::to_string(i++));
});
```

File: test/module/irohad/consensus/yac/yac_rainy_day_test.cpp [hunk 10]
```diff
@@ -153,19 +153,22 @@ TEST_F(YacTest, ValidCaseWhenReceiveOnVoteAfterReject) {
 
   std::vector<VoteMessage> votes;
   for (size_t i = 0; i < peers_number / 2; ++i) {
-    votes.push_back(
-        create_vote(hash1, my_order->getPeers().at(i).pubkey.to_string()));
+    auto peer = my_order->getPeers().at(i);
+    auto old_peer = *std::unique_ptr<iroha::model::Peer>(peer->makeOldModel());
```
Comment (l4l) [inline test/module/irohad/consensus/yac/yac_rainy_day_test.cpp]: Remove that line and use `crypto::toBinaryString(peer.pubkey().blob())` below

File: test/module/irohad/consensus/yac/yac_rainy_day_test.cpp [hunk 11]
```diff
@@ -153,19 +153,22 @@ TEST_F(YacTest, ValidCaseWhenReceiveOnVoteAfterReject) {
...
+    votes.push_back(create_vote(hash1, old_peer.pubkey.to_string()));
   };
   for (size_t i = peers_number / 2; i < peers_number - 1; ++i) {
-    votes.push_back(
-        create_vote(hash2, my_order->getPeers().at(i).pubkey.to_string()));
+    auto peer = my_order->getPeers().at(i);
+    auto old_peer = *std::unique_ptr<iroha::model::Peer>(peer->makeOldModel());
+    votes.push_back(create_vote(hash2, old_peer.pubkey.to_string()));
```
Comment (l4l) [inline test/module/irohad/consensus/yac/yac_rainy_day_test.cpp]: Same here
check the following on your own

File: test/module/irohad/validation/chain_validation_test.cpp [hunk 12]
```diff
@@ -54,9 +54,9 @@ class ChainValidationTest : public ::testing::Test {
   auto getBlockBuilder() const {
     return TestBlockBuilder()
         .transactions(std::vector<shared_model::proto::Transaction>{})
-        .txNumber(0)
         .height(1)
-        .prevHash(hash);
+        .prevHash(hash)
+        .createdTime(iroha::time::now());
```
Comment (l4l) [inline test/module/irohad/validation/chain_validation_test.cpp]: What's the purpose?

**Comment by neewy** [comment]
I am sorry, but how is it different from https://github.com/hyperledger/iroha/pull/1052? I think you are confusing reviewers, please tell how it happened

**Comment by x3medima17** [comment]
@neewy this branch was rebased onto `consensus-gate` instead of `develop`. That was an accident.
Now it is fixed

File: shared_model/backend/protobuf/block.hpp (hunk 4 above)
Comment (Alexey-N-Chernyshov) [reply shared_model/backend/protobuf/block.hpp]: move to Signable class

File: shared_model/backend/protobuf/queries/proto_query.hpp [hunk 13]
```diff
@@ -142,6 +142,11 @@ namespace shared_model {
         return true;
       }
 
+      bool clearSignatures() override {
```
Comment (Alexey-N-Chernyshov) [inline shared_model/backend/protobuf/queries/proto_query.hpp]: Query set has only one signature.
Make sure it is cleared - may be add test for it?

File: libs/common/byteutils.hpp [hunk 14]
```diff
@@ -23,7 +23,7 @@
 #include <string>
 
 #include "common/types.hpp"
-
+#include <iostream>
```
Comment (Alexey-N-Chernyshov) [inline libs/common/byteutils.hpp]: I think it should be removed

File: irohad/consensus/yac/impl/yac_gate_impl.cpp [hunk 15]
```diff
@@ -24,6 +25,12 @@
 #include "network/block_loader.hpp"
 #include "simulator/block_creator.hpp"
 
+#include "backend/protobuf/from_old_model.hpp"
```
Comment (Alexey-N-Chernyshov) [inline irohad/consensus/yac/impl/yac_gate_impl.cpp]: please, remove extra line and reorder includes

File: irohad/consensus/yac/impl/yac_gate_impl.cpp [hunk 16]
```diff
@@ -42,92 +49,109 @@ namespace iroha {
...
+          auto bl = shared_model::proto::from_old(block);
+          auto tmp = bl.prevHash();
+          this->vote(bl);
+        });
       }
 
-      void YacGateImpl::vote(model::Block block) {
-        auto hash = hash_provider_->makeHash(block);
+      void YacGateImpl::vote(const shared_model::interface::Block &block) {
+        std::cout<<block.signatures().size()<<std::endl;
```
Comment (Alexey-N-Chernyshov) [inline irohad/consensus/yac/impl/yac_gate_impl.cpp]: std::cout? Is it for debugging?

File: test/module/irohad/consensus/yac/yac_gate_test.cpp [hunk 17]
```diff
@@ -26,6 +26,8 @@
 #include "consensus/yac/impl/yac_gate_impl.hpp"
 #include "cryptography/hash.hpp"
 #include "framework/test_subscriber.hpp"
+#include "backend/protobuf/from_old_model.hpp"
```
Comment (Alexey-N-Chernyshov) [inline test/module/irohad/consensus/yac/yac_gate_test.cpp]: clang-format for ordrering

File: test/module/irohad/consensus/yac/yac_gate_test.cpp [hunk 18]
```diff
@@ -52,6 +54,10 @@ class YacGateTest : public ::testing::Test {
     commit_message = CommitMessage({message});
     expected_commit = rxcpp::observable<>::just(commit_message);
 
+    auto bytes = shared_model::proto::from_old(expected_block).hash().blob();
+    std::copy(bytes.begin(), bytes.end(), expected_block.hash.begin());
+
+
```
Comment (Alexey-N-Chernyshov) [inline test/module/irohad/consensus/yac/yac_gate_test.cpp]: extra line

File: test/module/irohad/consensus/yac/yac_gate_test.cpp [hunk 19]
```diff
@@ -112,8 +118,13 @@ TEST_F(YacGateTest, YacGateSubscriptionTest) {
 
   // verify that yac gate emit expected block
   auto gate_wrapper = make_test_subscriber<CallExact>(gate->on_commit(), 1);
-  gate_wrapper.subscribe(
-      [this](auto block) { ASSERT_EQ(block, expected_block); });
+  gate_wrapper.subscribe([this](auto block) {
```
Comment (Alexey-N-Chernyshov) [inline test/module/irohad/consensus/yac/yac_gate_test.cpp]: may be rework with shared_model?
I think expected_block should be shared_model.

File: test/module/irohad/consensus/yac/yac_gate_test.cpp [hunk 20]
```diff
@@ -112,8 +118,13 @@ TEST_F(YacGateTest, YacGateSubscriptionTest) {
...
+    auto bl = *std::unique_ptr<iroha::model::Block>(block->makeOldModel());
+    //std::cout<<block->toString()<<std::endl;
```
Comment (Alexey-N-Chernyshov) [inline test/module/irohad/consensus/yac/yac_gate_test.cpp]: dead code

File: test/module/irohad/consensus/yac/yac_gate_test.cpp [hunk 21]
```diff
@@ -184,13 +195,14 @@ TEST_F(YacGateTest, LoadBlockWhenDifferentCommit) {
               std::make_shared<shared_model::proto::Block>(
                   shared_model::proto::from_old(expected_block)))));
 
-  init();
+   init();
```
Comment (Alexey-N-Chernyshov) [inline test/module/irohad/consensus/yac/yac_gate_test.cpp]: clang-format

File: test/module/irohad/consensus/yac/yac_gate_test.cpp [hunk 22]
```diff
@@ -184,13 +195,14 @@ TEST_F(YacGateTest, LoadBlockWhenDifferentCommit) {
...
 
   // verify that yac gate emit expected block
   auto gate_wrapper = make_test_subscriber<CallExact>(gate->on_commit(), 1);
   gate_wrapper.subscribe([this](auto block) {
-    block.hash = expected_block.hash;
-    ASSERT_EQ(block, expected_block);
+    auto bl = *std::unique_ptr<iroha::model::Block>(block->makeOldModel());
+    bl.hash = expected_block.hash;
+    ASSERT_EQ(bl, expected_block);
```
Comment (Alexey-N-Chernyshov) [inline test/module/irohad/consensus/yac/yac_gate_test.cpp]: change expected_block to the new model instead of making old model

File: test/module/irohad/synchronizer/synchronizer_test.cpp [hunk 23]
```diff
@@ -72,7 +72,7 @@ MATCHER_P(NewBlockMatcher,
 TEST_F(SynchronizerTest, ValidWhenInitialized) {
   // synchronizer constructor => on_commit subscription called
   EXPECT_CALL(*consensus_gate, on_commit())
-      .WillOnce(Return(rxcpp::observable<>::empty<Block>()));
+      .WillOnce(Return(rxcpp::observable<>::empty<std::shared_ptr<shared_model::interface::Block>>()));
```
Comment (Alexey-N-Chernyshov) [inline test/module/irohad/synchronizer/synchronizer_test.cpp]: please, clang-format if haven't done it yet

File: irohad/consensus/yac/impl/yac_gate_impl.cpp (hunk 1 above)
Comment (x3medima17) [reply irohad/consensus/yac/impl/yac_gate_impl.cpp]: Debug artefacts

File: shared_model/backend/protobuf/transaction.hpp [hunk 24]
```diff
@@ -82,6 +82,11 @@ namespace shared_model {
         return true;
       }
 
+      bool clearSignatures() override {
```
[the remark this answers is not in the document]
Comment (x3medima17) [reply shared_model/backend/protobuf/transaction.hpp]: `Signable` is an interface, thus it should not know anything about signature storage implementation

File: irohad/validation/impl/chain_validator_impl.cpp (hunk 3 above)
Comment (x3medima17) [reply irohad/validation/impl/chain_validator_impl.cpp]: That's @Alexey-N-Chernyshov change

File: test/module/irohad/synchronizer/synchronizer_test.cpp (hunk 23 above)
Comment (x3medima17) [reply test/module/irohad/synchronizer/synchronizer_test.cpp]: it's formatted

File: shared_model/backend/protobuf/queries/proto_query.hpp (hunk 13 above)
Comment (x3medima17) [reply shared_model/backend/protobuf/queries/proto_query.hpp]: Not the scope of this PR

File: irohad/validation/impl/chain_validator_impl.cpp (hunk 3 above)
Comment (Alexey-N-Chernyshov) [reply irohad/validation/impl/chain_validator_impl.cpp]: please, revert this

File: irohad/consensus/yac/impl/yac_gate_impl.cpp [hunk 25]
```diff
@@ -42,28 +48,32 @@ namespace iroha {
...
+            [this](auto block) { this->vote(*block); });
       }
 
-      void YacGateImpl::vote(model::Block block) {
-        auto hash = hash_provider_->makeHash(block);
+      void YacGateImpl::vote(const shared_model::interface::Block &block) {
+        std::unique_ptr<model::Block> bl(block.makeOldModel());
+        auto hash = hash_provider_->makeHash(*bl);
         log_->info(
             "vote for block ({}, {})", hash.proposal_hash, hash.block_hash);
```
[the remark this answers is not in the document]
Comment (Alexey-N-Chernyshov) [reply irohad/consensus/yac/impl/yac_gate_impl.cpp]: so, just use block.hash()

File: irohad/consensus/yac/impl/yac_gate_impl.cpp [hunk 26]
```diff
@@ -42,28 +48,32 @@ namespace iroha {
...
       }
 
-      void YacGateImpl::vote(model::Block block) {
-        auto hash = hash_provider_->makeHash(block);
+      void YacGateImpl::vote(const shared_model::interface::Block &block) {
+        std::unique_ptr<model::Block> bl(block.makeOldModel());
+        auto hash = hash_provider_->makeHash(*bl);
         log_->info(
-            "vote for block ({}, {})", hash.proposal_hash, hash.block_hash);
+            "vote for block ({}, {})", hash.proposal_hash, block.hash().toString());
```
Comment (Alexey-N-Chernyshov) [inline irohad/consensus/yac/impl/yac_gate_impl.cpp]: Please clarify is it possible for proposal_hash to differ from block_hash. It seems that it is redundant output and it is enough to print only `block.hash()`. If it is so, you don't need old model conversion at all.

File: irohad/consensus/yac/impl/yac_gate_impl.cpp [hunk 27]
```diff
@@ -16,13 +16,19 @@
  */
 
 #include "consensus/yac/impl/yac_gate_impl.hpp"
+#include "backend/protobuf/block.hpp"
+#include "backend/protobuf/from_old_model.hpp"
```
Comment (Alexey-N-Chernyshov) [inline irohad/consensus/yac/impl/yac_gate_impl.cpp]: It seems file doesn't needed anymore.

File: irohad/consensus/yac/impl/yac_gate_impl.cpp (hunk 26 above)
Comment (Alexey-N-Chernyshov) [reply irohad/consensus/yac/impl/yac_gate_impl.cpp]: Also after that you can remove old block from the yac_gate_test.

Comment (x3medima17) [reply irohad/consensus/yac/impl/yac_gate_impl.cpp]: Will change it in the next PR

File: irohad/consensus/yac/impl/yac_gate_impl.cpp (hunk 27 above)
Comment (x3medima17) [reply irohad/consensus/yac/impl/yac_gate_impl.cpp]: Same as above

File: irohad/network/consensus_gate.hpp [hunk 28]
```diff
@@ -32,14 +38,15 @@ namespace iroha {
       /**
        * Providing data for consensus for voting
```
Comment (l4l) [inline irohad/network/consensus_gate.hpp]: Would be nice to update doc comments (e.g `@param` & `@return`)

File: shared_model/backend/protobuf/block.hpp (hunk 4 above)
Comment (l4l) [reply shared_model/backend/protobuf/block.hpp]: Comment may be useful then, it is not obvious
